File: Android/melody-runtime/src/main/java/com/melody/runtime/state/LocalState.kt
package com.melody.runtime.state

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import com.melody.core.schema.StateValue
import com.melody.runtime.engine.LuaValue

/**
 * Component-scoped local state, similar to ScreenState but scoped to a single component.
 * Port of iOS LocalState.swift.
 */
class LocalState {
    private val slots = mutableMapOf<String, MutableState<LuaValue>>()
    /** Mirror of values without triggering Compose tracking */
    private val rawValues = mutableMapOf<String, LuaValue>()

    fun slot(key: String): MutableState<LuaValue> {
        return slots.getOrPut(key) { mutableStateOf(LuaValue.Nil) }
    }

    fun update(key: String, value: LuaValue) {
        slot(key).value = value
        rawValues[key] = value
    }

    fun get(key: String): LuaValue = slot(key).value

    fun initialize(defaults: Map<String, StateValue>?) {
        defaults?.forEach { (key, sv) ->
            val lv = LuaValue.fromStateValue(sv)
            slot(key).value = lv
            rawValues[key] = lv
        }
    }

    /** All values (triggers Compose tracking on every slot) */
    val allValues: Map<String, LuaValue>
        get() = slots.mapValues { it.value.value }

    /** All values without triggering Compose tracking */
    val allValuesUntracked: Map<String, LuaValue>
        get() = rawValues.toMap()
}

File: Sources/Runtime/State/LocalState.swift
import Foundation
import SwiftUI
import Core

// MARK: - Environment Key

private struct LocalStateKey: EnvironmentKey {
    static let defaultValue: LocalState? = nil
}

extension EnvironmentValues {
    var localState: LocalState? {
        get { self[LocalStateKey.self] }
        set { self[LocalStateKey.self] = newValue }
    }
}

// MARK: - Local State (NOT @Observable — observation is per-slot, reuses StateSlot)

/// Scoped state container for component-local variables, bridged to the Lua `scope` table.
@MainActor
final class LocalState {
    private var slots: [String: StateSlot] = [:]
    /// Mirror of slot values that can be read without triggering @Observable tracking.
    /// Kept in sync by update() and initialize().
    private var rawValues: [String: LuaValue] = [:]

    nonisolated init() {}

    func slot(for key: String) -> StateSlot {
        if let existing = slots[key] { return existing }
        let new = StateSlot()
        slots[key] = new
        return new
    }

    func update(key: String, value: LuaValue) {
        slot(for: key).value = value
        rawValues[key] = value
    }

    func get(key: String) -> LuaValue {
        slot(for: key).value
    }

    func initialize(from defaults: [String: StateValue]?) {
        guard let defaults else { return }
        for (key, sv) in defaults {
            let lv = sv.toLuaValue()
            slot(for: key).value = lv
            rawValues[key] = lv
        }
    }

    /// All values — reads through @Observable (registers tracking on every slot).
    var allValues: [String: LuaValue] {
        slots.mapValues { $0.value }
    }

    /// All values without triggering @Observable tracking.
    /// Use this when pushing scope to Lua during body evaluation.
    var allValuesUntracked: [String: LuaValue] {
        rawValues
    }
}
